ngraph/test/layout.cpp

```diff
@@ -0,0 +1,197 @@
+//
+
+#include "openvino/core/layout.hpp"
+
+#include "gtest/gtest.h"
+
+using namespace ov;
+
+TEST(layout, basic) {
+    Layout l = "NcDHw";
+    EXPECT_TRUE(layouts::has_batch(l));
+    EXPECT_EQ(layouts::batch(l), 0);
+    EXPECT_TRUE(layouts::has_channels(l));
+    EXPECT_EQ(layouts::channels(l), 1);
+    EXPECT_TRUE(layouts::has_depth(l));
+    EXPECT_EQ(layouts::depth(l), 2);
+    EXPECT_TRUE(layouts::has_height(l));
+    EXPECT_EQ(layouts::height(l), 3);
+    EXPECT_TRUE(layouts::has_width(l));
+    EXPECT_EQ(layouts::width(l), 4);
+}
+
+TEST(layout, advanced_syntax) {
+    Layout l = "[batch, channels, depth, height, width]";
+    EXPECT_TRUE(layouts::has_batch(l));
+    EXPECT_EQ(layouts::batch(l), 0);
+    EXPECT_TRUE(layouts::has_channels(l));
+    EXPECT_EQ(layouts::channels(l), 1);
+    EXPECT_TRUE(layouts::has_depth(l));
+    EXPECT_EQ(layouts::depth(l), 2);
+    EXPECT_TRUE(layouts::has_height(l));
+    EXPECT_EQ(layouts::height(l), 3);
+    EXPECT_TRUE(layouts::has_width(l));
+    EXPECT_EQ(layouts::width(l), 4);
+    EXPECT_EQ(l, Layout("ncdhw"));
+
+    l = "[custom1, ?, custom2]";
+    EXPECT_EQ(l.get_index_by_name("CUSTOm1"), 0);
+    EXPECT_EQ(l.get_index_by_name("Custom2"), 2);
+
+    l = "[?, N , ... , channels , ?]";
+    EXPECT_EQ(l, Layout("?N...C?"));
+    l = "[?, N , ... , custom1 , ?]";
+    EXPECT_EQ(l.get_index_by_name("Custom1"), -2);
+}
+
+TEST(layout, empty) {
+    Layout l;
+    EXPECT_FALSE(layouts::has_batch(l));
+    EXPECT_ANY_THROW(layouts::batch(l));
+    EXPECT_FALSE(layouts::has_channels(l));
+    EXPECT_ANY_THROW(layouts::channels(l));
+    EXPECT_FALSE(layouts::has_depth(l));
+    EXPECT_ANY_THROW(layouts::depth(l));
+    EXPECT_FALSE(layouts::has_height(l));
+    EXPECT_ANY_THROW(layouts::height(l));
+    EXPECT_FALSE(layouts::has_width(l));
+    EXPECT_ANY_THROW(layouts::width(l));
+}
+
+TEST(layout, to_string) {
+    std::vector<Layout> layouts = {{"NCHW"},
+                                   {"[?, N, CHANNELS, ?, ?, Custom_dim_name]"},
+                                   {"012?...3?456"},
+                                   {"...3?456"},
+                                   {"12?34..."},
+                                   {"..."},
+                                   Layout::scalar()};
+    for (const auto& l : layouts) {
+        EXPECT_EQ(l, Layout(l.to_string()));
+    }
+}
+
+TEST(layout, scalar) {
+    auto l = Layout::scalar();
+    EXPECT_FALSE(layouts::has_batch(l));
+    EXPECT_ANY_THROW(layouts::batch(l));
+    EXPECT_FALSE(layouts::has_channels(l));
+    EXPECT_ANY_THROW(layouts::channels(l));
+    EXPECT_FALSE(layouts::has_depth(l));
+    EXPECT_ANY_THROW(layouts::depth(l));
+    EXPECT_FALSE(layouts::has_height(l));
+    EXPECT_ANY_THROW(layouts::height(l));
+    EXPECT_FALSE(layouts::has_width(l));
+    EXPECT_ANY_THROW(layouts::width(l));
+}
+
+TEST(layout, custom_dims) {
+    Layout l = "0ac";
+    EXPECT_FALSE(layouts::has_batch(l));
+    EXPECT_ANY_THROW(layouts::batch(l));
```

Comment:
Why any throw here?

Reply:
Ok, will correct. Also will change exception type from ngraph_error to ov::AssertFailure